privacyidea/lib/container.py

```diff
@@ -0,0 +1,782 @@
+            prev = page - 1
+        nxt = None
+        if pagination.has_next:
+            nxt = page + 1
+
+        ret["prev"] = prev
+        ret["next"] = nxt
+        ret["current"] = page
+        ret["count"] = pagination.total
+    else:  # No pagination
+        db_containers = sql_query.all()
+
+    container_list = [create_container_from_db_object(db_container) for db_container in db_containers]
+    ret["containers"] = container_list
+
+    return ret
+
+
+def find_container_for_token(serial):
+    """
+    Returns a TokenContainerClass object for the given token or raises a ResourceNotFoundError
+    if the token does not exist.
+
+    :param serial: Serial of the token
+    :return: container object or None if the token is not in a container
+    """
+    container = None
+    db_token = Token.query.filter(Token.serial == serial).first()
+    if not db_token:
+        raise ResourceNotFoundError(f"Unable to find token with serial {serial}.")
+    token_id = db_token.id
+    row = TokenContainerToken.query.filter(TokenContainerToken.token_id == token_id).first()
+    if row:
+        container_id = row.container_id
+        container = find_container_by_id(container_id)
+    return container
+
+
+def get_container_classes():
+    """
+    Returns a dictionary of all available container classes in the format: { type: class }.
+    New container types have to be added here.
+    """
+    # className: module
+    classes = {
+        "TokenContainerClass": "privacyidea.lib.containerclass",
+        "SmartphoneContainer": "privacyidea.lib.containers.smartphone",
+        "YubikeyContainer": "privacyidea.lib.containers.yubikey"
+    }
+
+    ret = {}
+    for cls, mod in classes.items():
+        try:
+            m = importlib.import_module(mod)
+            c = getattr(m, cls)
+            ret[c.get_class_type().lower()] = c
+        except Exception as ex:  # pragma: no cover
+            log.error(f"Error importing module {cls}: {ex}")
+
+    return ret
+
+
+def get_container_policy_info(container_type=None):
+    """
+    Returns the policy info for the given container type or for all container types if no type is defined.
+
+    :param container_type: The type of the container, optional
+    :return: The policy info for the given container type or for all container types
+    """
+    classes = get_container_classes()
+    if container_type:
+        if container_type in classes.keys():
+            return classes[container_type].get_container_policy_info()
+        else:
+            raise ResourceNotFoundError(f"Unable to find container type {container_type}.")
+    else:
+        ret = {}
+        for container_type, container_class in classes.items():
+            ret[container_type] = container_class.get_container_policy_info()
+        return ret
+
+
+def init_container(params):
+    """
+    Create a new container with the given parameters. Requires at least the type.
+
+    :jsonparam type: The type of the container
```

Comment:
since this is no API function, `jsonparam` doesn't really work here. We should describe these as required contents of the `params` dictionary.